docs: Add contributing guidelines

mgalka opened the pull request:
Add a few guidelines about filing PRs agains kernelci-core.

Signed-off-by: Michal Galka <michal.galka@collabora.com>

mgalka commented:
This PR is gets the content of https://github.com/kernelci/kernelci-core/pull/1562 to the top level menu of the docs.

@aliceinwire @gctucker @broonie @khilman Is it ok to merge it?

nuclearcat reviewed (approved):
LGTM

ojeda commented on kernelci.org/content/en/docs/contrib.md:
```diff
@@ -0,0 +1,33 @@
+---
+
+KernelCI core project is open for contributions. Contributions may consist of
+adding new build, tests and device types as well as features and bugfixes for
```
Should it be plural?

```suggestion
adding new builds, tests and device types as well as features and bugfixes for
```

ojeda commented on kernelci.org/content/en/docs/contrib.md:
```diff
@@ -0,0 +1,33 @@
+adding new build, tests and device types as well as features and bugfixes for
+KernelCI core tools.
+The best way to contribute is to send a PR to [kernelci-core](https://github.com/kernelci/kernelci-core).
+When the PR is created, [KernelCI staging](https://kernelci.org/docs/instances/staging)
```
```suggestion
When the PR is created, the [KernelCI staging](https://kernelci.org/docs/instances/staging)
```

ojeda commented on kernelci.org/content/en/docs/contrib.md:
```diff
@@ -0,0 +1,33 @@
+When the PR is created, [KernelCI staging](https://kernelci.org/docs/instances/staging)
+instance takes care of updating the [staging.kernelci.org branch](https://github.com/kernelci/kernelci-core/tree/staging.kernelci.org).
+In general the branch is updated every 8h and a limited set of builds and tests
+is run on it. More detailed information about the logic behind staging runs can
```
```suggestion
are run on it. More detailed information about the logic behind staging runs can
```

mgalka replied:
I'm not sure if it should be **are** I'm mentioning a _set_ here, that's why I used **is**.  Having said that I'm not a native English speaker so I might be wrong. @gctucker maybe you could help?

gctucker replied:
Even if the noun is singular, when it designates a plural entity it should be treated as plural e.g. "the team are on track to win the championship".  So it should be "are" imho.

ojeda replied:
I think either is fine, as far as I understand it depends on what one wants to emphasize, and singular is more common in American English (https://en.wikipedia.org/wiki/American_and_British_English_grammatical_differences#Subject-verb_agreement).

ojeda commented on kernelci.org/content/en/docs/contrib.md:
```diff
@@ -0,0 +1,33 @@
+instance takes care of updating the [staging.kernelci.org branch](https://github.com/kernelci/kernelci-core/tree/staging.kernelci.org).
+In general the branch is updated every 8h and a limited set of builds and tests
+is run on it. More detailed information about the logic behind staging runs can
+be found [here](https://kernelci.org/docs/instances/staging).
```
```suggestion
be found in the [staging instance documentation](https://kernelci.org/docs/instances/staging).
```

gctucker replied:
I think `here` links are bad, the text should be informative regardless of any hyperlinks.  That's some web design "best practice" rule I learnt a while ago...  So I'd rather leave `the staging instance documentation`.

ojeda replied:
Yeah, exactly (though I am not sure if you are agreeing or perhaps you read the diff the other direction).

gctucker replied:
Ah yes sorry, my bad :coffee:

ojeda commented on kernelci.org/content/en/docs/contrib.md:
```diff
@@ -0,0 +1,33 @@
+is run on it. More detailed information about the logic behind staging runs can
+be found [here](https://kernelci.org/docs/instances/staging).
+
+There are several guidelines which can facilitate the PR review process.
```
```suggestion
There are several guidelines which can facilitate the PR review process:
```

ojeda commented on kernelci.org/content/en/docs/contrib.md:
```diff
@@ -0,0 +1,33 @@
+   1. Describe the purpose of the changes
+   2. Example use cases are welcome
+2. Attach staging build/test results when possible
+   1. Staging instance is updated every 8h. If the PR is expected to produce build/test results
+   check [staging dashboard](https://staging.kernelci.org) and make sure these are mentioned in the PR comment
```
Perhaps an example of how the ideal PR comments and what sort of link to staging is expected would be useful. In particular, the links may help to understand the kind of URLs/paths expected.

Also:

```suggestion
   1. The staging instance is updated every 8h. If the PR is expected to produce build/test results, then
   check [staging dashboard](https://staging.kernelci.org) and make sure these are mentioned in the PR comment
```

mgalka replied:
Actually, I looked for a PR that could be put here as an example, but all I looked into were using links to logs on stored on staging. I decided to not put it here as it was agreed that we're going to either paste log excerpts to the comments or feed data to  pastebin and paste links from there.

ojeda commented on kernelci.org/content/en/docs/contrib.md:
```diff
@@ -0,0 +1,33 @@
+3. Make sure that reviewers' comments and questions are addressed
+   1. When there are comments unanswered for more than 1 month the PR will be closed
+4. In case there is a need to consult the PR with KernelCI maintainers join the open hours
+   1. Open hours take place every Thursday at 12:00 UTC at KernelCI [Jitsi](https://meet.kernel.social/kernelci-dev)
+5. You can reach KernelCI maintainers also at:
+   1. IRC - #kernelci channel on [libera.chat](https://libera.chat/)
+   2. KernelCI mailing list - [kernelci@lists.linux.dev](mailto:kernelci@lists.linux.dev)
```
To avoid duplication, perhaps it is best to link to a page containing these contact details?

gctucker replied:
Namely `/docs/org/maintainers/`.

ojeda commented on kernelci.org/content/en/docs/contrib.md:
```diff
@@ -0,0 +1,33 @@
+   1. Describe the purpose of the changes
+   2. Example use cases are welcome
+2. Attach staging build/test results when possible
+   1. Staging instance is updated every 8h. If the PR is expected to produce build/test results
```
The first sentence is already explained above -- should it be removed here?

ojeda replied:
The GitHub UI makes my comment hard to understand -- I am referring here to line 24 but GitHub shows context, i.e. in particular I was talking about the "Staging instance is updated every 8h." sentence.

mgalka replied:
I put it there as a reminder, just to make it clear that the author needs to wait at least 8h to see results on staging. On second thought, It may be too much.

ojeda replied:
Yeah, I think either way is fine -- I didn't know if it was intentional. In some types of material, there is value on repeating key information/concepts.

One alternative is saying something shorter that reminds the reader about it, without duplicating information. For instance, something along the lines of "If build/test results are expected to be visible, then link them from staging." (which also allows you to just say "If not" in the next line, if I understand correctly)

aliceinwire commented on kernelci.org/content/en/docs/contrib.md:
```diff
@@ -0,0 +1,33 @@
+1. Make sure the PR is well described
+   1. Describe the purpose of the changes
+   2. Example use cases are welcome
+2. Attach staging build/test results when possible
```
Use a permanent link for staging build/test results when possible, as current links get removed after few weeks.
maybe suggest pastebin ( like github gist )

nuclearcat replied:
IMHO it is a bit hard to pastebin results from dashboard, where we have heavy ajax. Probably it is acceptable to have expiring link for new tests PR and just as confirmation that all works as designed.
There is very small chance we will need to review this test results.
But i agree that we need permanent links in case of discovered real bugs or commits.

aliceinwire replied:
> where we have heavy ajax.
I'm talking mostly about logs that are on https://storage.staging.kernelci.org/ and https://storage.kernelci.org/ that are temporary

aliceinwire replied:
just use something like wgetpastebin

gctucker replied:
Staging builds and tests aren't permanent.

Things we can treat as permanent are mailing list archives, anything on GitHub (issues, PRs etc.) and in theory results on the production dashboard although that's more a grey area as they might disappear after a while too.

I think it's best if we can just mention some log excerpts in the commit messages directly inline, and use `Link:` for email threads like in the kernel git history.

aliceinwire commented on kernelci.org/content/en/docs/contrib.md:
```diff
@@ -0,0 +1,33 @@
+
+There are several guidelines which can facilitate the PR review process.
+
+1. Make sure the PR is well described
```
We are missing a explanation about KernelCI usage of DCO (Signed-off-by: reviewed-by: tested-by:) in comments.

mgalka replied:
Is it OK to add it in as a follow-up PR?

aliceinwire replied:
ok

gctucker replied:
@mgalka Maybe create a GitHub issue so we don't forget about this by merging this PR?

gctucker replied:
Have you created an issue for this?

gctucker replied:
@mgalka See above - I haven't found any issue. I think that's the only thing to resolve before it can be merged

mgalka commented:
I pushed an updated version of contributing guidelines in which I tried to addressed the comments that appeared during the review. @gctucker @ojeda @aliceinwire Could you have another look at it? If it's OK for you, I think we can land it.

mgalka commented:
@gctucker @aliceinwire @ojeda 
If there are no more comments I'll merge the PR today so it can get to the production on Monday (13th of Feb).
Currently, you can see it on staging https://static.staging.kernelci.org/docs/contrib/

nuclearcat reviewed (approved):
great, thanks for your efforts!
